network/src/peers/peers.rs

```diff
@@ -165,44 +182,37 @@ impl<S: Storage + Send + Sync + 'static> Node<S> {
     ///
     /// Broadcasts a connection request to all disconnected peers.
     ///
-    async fn connect_to_disconnected_peers(&self) {
+    async fn connect_to_disconnected_peers(&self, count: usize) {
         // Local address must be known by now.
         let own_address = self.local_address().unwrap();
 
-        // If this node is a bootnode, attempt to connect to all disconnected peers.
         // If this node is not a bootnode, attempt to satisfy the minimum number of peer connections.
         let random_peers = {
-            // Fetch the number of connected and connecting peers.
-            let number_of_peers = self.peer_book.get_active_peer_count() as usize;
-
-            // Check if this node server is below the permitted number of connected peers.
-            let min_peers = self.config.minimum_number_of_connected_peers() as usize;
-            if number_of_peers >= min_peers {
-                return;
-            }
-
-            // Set the number of peers to attempt a connection to.
-            let count = min_peers - number_of_peers;
-
-            if count == 0 {
-                return;
-            }
-
-            let disconnected_peers = self.peer_book.disconnected_peers();
-
             trace!(
                 "Connecting to {} disconnected peers",
-                cmp::min(count, disconnected_peers.len())
+                cmp::min(count, self.peer_book.disconnected_peers().len())
             );
 
             let bootnodes = self.config.bootnodes();
 
             // Iterate through a selection of random peers and attempt to connect.
-            disconnected_peers
-                .iter()
-                .filter(|peer| **peer != own_address && !bootnodes.contains(peer))
-                .copied()
-                .choose_multiple(&mut rand::thread_rng(), count)
+            let mut candidates = self.peer_book.disconnected_peers_snapshot().await;
+
+            candidates.retain(|peer| peer.address != own_address && !bootnodes.contains(&peer.address));
+
+            if self.config.is_bootnode() {
+                // Bootnodes choose peers they haven't dialed in a while.
```

Comment:
this is a good idea, but it must be paired with a mechanism for discarding or skipping inaccessible peers from the list of disconnected peers, otherwise the bootnodes might only try to connect to old peers that might never become online

Reply:
This is already somewhat the case with the introduction of `judge_bad` in lockless. Before rebasing/rebuilding, I was tracking `is_routable` as a property set in case of connection failure, left this out for now so it could be discussed.

Reply:
Right, `judge_bad_offline` indeed mitigates this issue :+1:.